circuit_knitting_toolbox/circuit_cutting/cutting_evaluation.py

```diff
@@ -77,8 +200,252 @@ def append_measurement_circuit(
+        # with QuantumCircuits in some versions. (e.g. passes local pytest but fails in tox env)
+        quasi_dists_reshaped = np.reshape(
+            quasi_dists_flat, (len(subexperiments), len(subexperiments[0]))
+        )
+        num_qpd_bits = np.reshape(
+            num_qpd_bits_flat, (len(subexperiments), len(subexperiments[0]))
+        )
+
+    # Create the counts tuples, which include the number of QPD measurement bits
+    quasi_dists: list[list[tuple[dict[str, float], int]]] = [
+        [] for _ in range(len(subexperiments))
+    ]
+    for i, sample in enumerate(quasi_dists_reshaped):
+        for j, prob_dict in enumerate(sample):
+            quasi_dists[i].append((prob_dict, num_qpd_bits[i][j]))
+
+    return quasi_dists
+
+
+def _get_mapping_ids_by_partition(
+    circuits: Sequence[QuantumCircuit],
+) -> tuple[list[list[list[int]]], list[list[int]]]:
+    """Get indices to the QPD gates in each subcircuit and relevant map ids."""
+    # Collect QPDGate id's and relevant map id's for each subcircuit
+    subcirc_qpd_gate_ids: list[list[list[int]]] = []
+    subcirc_map_ids: list[list[int]] = []
+    decomp_ids = set()
+    for circ in circuits:
+        subcirc_qpd_gate_ids.append([])
+        subcirc_map_ids.append([])
+        for i, inst in enumerate(circ.data):
+            if isinstance(inst.operation, SingleQubitQPDGate):
+                decomp_id = int(inst.operation.label.split("_")[-1])
+                decomp_ids.add(decomp_id)
+                subcirc_qpd_gate_ids[-1].append([i])
+                subcirc_map_ids[-1].append(decomp_id)
+
+    return subcirc_qpd_gate_ids, subcirc_map_ids
+
+
+def _get_bases_by_partition(
+    circuits: Sequence[QuantumCircuit], subcirc_qpd_gate_ids: list[list[list[int]]]
+) -> list[QPDBasis]:
+    """Get a list of each unique QPD basis across the subcircuits."""
+    # Collect the bases corresponding to each decomposed operation
+    bases_dict = {}
+    for i, subcirc in enumerate(subcirc_qpd_gate_ids):
+        for basis_id in subcirc:
+            decomp_id = int(
+                circuits[i].data[basis_id[0]].operation.label.split("_")[-1]
+            )
+            bases_dict[decomp_id] = circuits[i].data[basis_id[0]].operation.basis
+    bases = [bases_dict[key] for key in sorted(bases_dict.keys())]
+
+    return bases
+
+
+def _get_bases(circuit: QuantumCircuit) -> tuple[list[QPDBasis], list[list[int]]]:
+    """Get a list of each unique QPD basis in the circuit and the QPDGate indices."""
+    bases = []
+    qpd_gate_ids = []
+    for i, inst in enumerate(circuit):
+        if isinstance(inst.operation, SingleQubitQPDGate):
+            raise ValueError(
+                "SingleQubitQPDGates are not supported in unseparable circuits."
+            )
+        if isinstance(inst.operation, TwoQubitQPDGate):
+            bases.append(inst.operation.basis)
+            qpd_gate_ids.append([i])
+
+    return bases, qpd_gate_ids
+
+
+def _validate_samplers(
```

Comment:
As I think about this more, I am becoming concerned about this business of intercepting the sampler and replacing it with a sampler that is different than the one the user provided.  If this going to be done, it should be done in a way that the user is informed what is happening, but this is difficult for two reasons:

1. A warning will only be displayed once in a given Python session
2. In the notebooks, we often use `%%capture` in cells that call the samplers, so the warning never makes it to the user.

I think it is far better to try to use the provided sampler and to fail with an informative error message, if possible, if the sampler is not suitable for our us.  We could create a small "canary" circuit, which we run if provided a Terra or Aer sampler.  If we run this circuit and it either fails or gives an unexpected result, we inform the user what failed and provide a suggestion.  Otherwise, we can go forward assuming the user is now on a more recent version of Terra or Aer, where the issues we've found with these samplers have been fixed.

If we want, for user convenience, we could even have a `def sampler_factory(*, shots=1024)` function in `simulation.py`, which would return an appropriately configuredAer sampler if `shots is not None` and `ExactSampler` otherwise.  In the future, when Aer's sampler supports everything we need it to, we can deprecate this and suggest people use Aer's sampler directly instead.  This convenience function would exist to allow the user to adjust one parameter (`shots`) and get the most appropriate sampler in every situation.

Overall, I think it is much better for us to have a dialogue with the user than do something that is beyond their control and likely beyond their awareness.  If a user is investigating something unexpected, they might try with different samplers, but if we're swapping out the sampler behind the scenes, this is wasted effort.  Further, being honest about the limitations of current samplers is probably good for us: it raises awareness of these issues rather than sweeping them under the rug.

Reply:
OK, let's just raise a `ValueError` and tell users to use `circuit_knitting_toolbox.utils.ExactSampler` if they want exact results.

Then the question is, what do we use in the tutorials? I like the exact expvals in tutorials. If we're going to have undeterministic results in tutorials, I'd rather run a bunch of experiments and show how the results are scattered around the true mean or something.

Reply:
> let's just raise a `ValueError`

I think something like `RuntimeError` is more appropriate, but I am interested to hear your reasoning.

> Then the question is, what do we use in the tutorials?

Yeah, that was in part the motivation behind my `sampler_factory` function.  Then we could use `sampler_factory(shots=None)` in the tutorials, and if the user actually wanted to have a finite number of shots, they could adjust it right there and then the aer sampler would be used instead.

Reply:
I think I'd rather just error out if the user passes an exact sampler that doesnt support midcirc meas. We can inform them exactly why their sampler doesnt work and how they can fix it (user ExactSampler).

Like you said, ExactSampler is non-physical, and I see it more of a debugging/learning tool. I think we should continue allowing users to directly set up their Sampler and pass it in. If we do the factory, it's going to need runtime info and options be passed in, and I'm sure there is other complexity. I'd rather not introduce that as a step.

Reply:
> If we do the factory, it's going to need runtime info and options be passed in, and I'm sure there is other complexity. I'd rather not introduce that as a step.

If we do the factory, I was thinking it should expose a very simple API, where `shots` is the only thing you can vary.  If you need anything more complex than that, use the appropriate class directly.

That said, I'm not certain this would be more clear if the class is hidden out in `simulation.py`.  In fact, maybe the best way to do this is to define it **directly in the notebook**, rather than in `simulation.py`:

```python
def best_sampler(shots: int = 1024) -> BaseSampler:
    if shots is None:
        return ExactSampler()
    else:
        return AerSampler(shots=shots)  # or whatever is necessary to set the shots count correctly
```

Reply:
OK let's do this in the notebook and write a little comment specifying why it is necessary